TRC selection: remove python code (#133)
scionassociation/scion-cppki_I-D: draft-dekater-scion-pki.md

- The REQUIRED signatures for new certificates are implied by the currently valid TRC payload, and, in case of a TRC update, the predecessor payload.

## Certification Path - Trust Anchor Pool {#trc-selection}

The certification path of a Control Plane AS certificate starts in a Control Plane root certificate. While root certificates for a given ISD are distributed via the TRC, AS and issuing CA certificates are distributed separately. This separation makes it possible to extend the validity period of the root certificate, and to update the corresponding TRC without having to modify the certificate chain.

> To validate a certification path, a relying party builds a collection of root certificates know as the trust anchor pool. Because TRC updates can introduce a grace period where multiple TRCs overlap, relying parties MUST execute the following steps to determine the correct trust anchor pool for a given verification time:

Reviewer: ```suggestion
To validate a certification path, a relying party builds a collection of root certificates known as the trust anchor pool. Because TRC updates can introduce a grace period where multiple TRCs overlap, relying parties MUST execute the following steps to determine the correct trust anchor pool for a given verification time:
```
Typo?

> 1. From the set of all available TRCs for the ISD, filter out TRCs whose validity start time (`notBefore` date) precedes the verification time.

Reviewer: ```suggestion
1. From the set of all available TRCs for the ISD, keep only TRCs whose validity start time (`notBefore` date) is at or before the verification time.
```

IIUC, this effectively says "_remove_ (filter out) TRCs whose notBefore _precedes_ the verification time", but those are the TRCs that became eligible for selection, no?

Or maybe it's "filter out" that's ambiguous and potentially misleading.

> 1. From the set of all available TRCs for the ISD, keep only TRCs whose validity start time (`notBefore` date) is at or before the verification time.

Reviewer: ```suggestion
1. From the set of all available TRCs for the ISD, keep only TRCs whose validity start time (`notBefore` date) is at or before the verification time. If no such TRC exists, the process terminates unsuccessfully.
```

> 2. From the remaining TRCs, identify those with the highest base number, then select the TRC with the highest serial number.

Reviewer: ```suggestion
2. From the selected TRCs, identify those with the highest base number (`baseNumber`), then select the TRC among them with the highest serial number (`serialNumber`).
```

> 3. If the verification time is strictly greater than the TRC's `notAfter` date, it is no longer valid. There are no valid trust anchors at the verification time, and the process terminates.

Reviewer: ```suggestion
3. If the verification time is strictly greater than the selected TRC's `notAfter` date then the process terminates unsuccessfully.
```

The selection of the right set of TRCs to build the trust anchor pool depends on the time of verification. The trust anchor pool is usually used to verify control plane messages and in this case, the time of verification is the current time. However, if the trust anchor pool will be used for auditing, the time of verification is the point in time to check whether a given signature was verifiable.

To construct the trust anchor pool for a specific ISD at a given `verification_time`, a relying party executes the following steps:

1. Determine the Active Base Number: Filter the set of all available TRCs for the ISD to those whose `notBefore` validity date is less than or equal to the `verification_time`. From this filtered set, select the highest `baseNumber`.

2. Identify the Candidate TRC: Filter the set of available TRCs again to find only those that match the `baseNumber` identified in Step 1, and whose `notBefore` date is less than or equal to the `verification_time`. From this set, select the TRC with the highest `serialNumber`. This is the *Candidate TRC*.

> 3. Verify Candidate TRC Expiration:**

Reviewer: what are those trailing `**`?

   If the `verification_time` is strictly greater than the Candidate TRC's `notAfter` date, the algorithm terminates and returns an empty set. There are no valid trust anchors for this time.

4. Extract Candidate Trust Anchors:**
   Parse the `certificates` sequence of the Candidate TRC. Extract all certificates that possess a `basicConstraints` extension with the `cA` boolean asserted. Add these CP Root Certificates to the trust anchor pool.

5. Evaluate the Grace Period:**
   Check if the Candidate TRC is currently within its grace period (i.e., `verification_time` <= `notBefore` + `gracePeriod`). If the grace period has expired, the algorithm terminates and returns the current trust anchor pool.

6. Process the Predecessor TRC (If Applicable):

>   If the Candidate TRC is still within its grace period, attempt to retrieve the Predecessor TRC (the TRC with the same `baseNumber` and a `serialNumber` exactly one less than the Candidate TRC).

Reviewer: Missing space here?

```suggestion
   If the Candidate TRC is still within its grace period, attempt to retrieve the Predecessor TRC (the TRC with the same `baseNumber` and a `serialNumber` exactly one less than the Candidate TRC).
```
